src/Middleware/WebSockets/src/WebSocketMiddleware.cs

```diff
@@ -139,21 +144,57 @@ public async Task<WebSocket> AcceptAsync(WebSocketAcceptContext acceptContext)
                 }
 
                 TimeSpan keepAliveInterval = _options.KeepAliveInterval;
+                bool enableCompression = false;
+                bool serverContextTakeover = true;
+                int serverMaxWindowBits = 15;
                 if (acceptContext is ExtendedWebSocketAcceptContext advancedAcceptContext)
                 {
                     if (advancedAcceptContext.KeepAliveInterval.HasValue)
                     {
                         keepAliveInterval = advancedAcceptContext.KeepAliveInterval.Value;
                     }
+                    enableCompression = advancedAcceptContext.DangerousEnableCompression;
+                    serverContextTakeover = !advancedAcceptContext.DisableServerContextTakeover;
+                    serverMaxWindowBits = advancedAcceptContext.ServerMaxWindowBits;
                 }
 
                 string key = _context.Request.Headers.SecWebSocketKey;
 
                 HandshakeHelpers.GenerateResponseHeaders(key, subProtocol, _context.Response.Headers);
 
+                WebSocketDeflateOptions? deflateOptions = null;
+                if (enableCompression)
+                {
+                    var ext = _context.Request.Headers.SecWebSocketExtensions;
+                    if (ext.Count != 0)
+                    {
+                        // loop over each extension offer, extensions can have multiple offers, we can accept any
+                        foreach (var extension in _context.Request.Headers.GetCommaSeparatedValues(HeaderNames.SecWebSocketExtensions))
+                        {
+                            if (extension.AsSpan().TrimStart().StartsWith("permessage-deflate", StringComparison.Ordinal))
+                            {
+                                if (HandshakeHelpers.ParseDeflateOptions(extension.AsSpan().TrimStart(), serverContextTakeover, serverMaxWindowBits, out var parsedOptions, out var response))
```

Comment:
We'll want some debug logging around the success or failure of this negotiation and the resulting values that will be used.